### PR TITLE
Replace `zip_choices` with strict `zip`

### PR DESCRIPTION
**Pull request by Kakadus**
Another commit extracted for #2051. The `zip_choices` filter zipped choices' `counts` and `names`. This asserts, that the no answer option is always the last option.

### CODE REVIEW

File: evap/results/tools.py [hunk 1]
```diff
@@ -48,10 +48,14 @@ def __init__(self, question, answer_counters, additional_text_result=None):
...
+            counts = OrderedDict(
+                (value, [0, name, color, value]) for (name, color, value) in self.choices.as_name_color_value_tuples()
+            )
+            counts.pop(NO_ANSWER)
             for answer_counter in answer_counters:
-                counts[answer_counter.answer] = answer_counter.count
-            self.counts = tuple(counts.values())
+                counts[answer_counter.answer][0] = answer_counter.count
+            self.zipped_choices = tuple(counts.values())
+            self.counts = tuple((count for count, _, _, _ in counts.values()))
```
Comment (niklasmohrin) [inline evap/results/tools.py:58]: This feels overly complicated, maybe something like the following is better?
```python
            answer_indices = {
                answer: i for i, answer in enumerate(self.choices.values) if answer != NO_ANSWER
            }
            counts = [0] * len(answer_indices)
            for counter in answer_counters:
                counts[answer_indices[counter.answer]] = counter.count
            self.counts = tuple(counts)
            self.zipped_choices = self.choices.as_name_color_value_tuples()
```
We should probably refactor the whole choices business to use dataclasses instead of named tuples and increase typing with these result classes here, but we don't need to do it now

Comment (Kakadus) [reply evap/results/tools.py]: The problem is, that `self.choices.as_name_color_value_tuples()` contains a `NO_ANSWER`-element as the last element. This needs to be removed from the result of `as_name_color_value_tuples`. Also, the counts need to be prepended

File: evap/evaluation/templatetags/evaluation_filters.py [hunk 2]
```diff
@@ -115,12 +111,12 @@ def percentage_one_decimal(fraction, population):
...
 @register.filter
-def to_colors(choices):
-    if not choices:
+def to_colors(question_result: RatingResult | None):
+    if not question_result:
         # When displaying the course distribution, there are no associated voting choices.
         # In that case, we just use the colors of a unipolar scale.
-        return BASE_UNIPOLAR_CHOICES["colors"]
-    return choices.colors
+        return BASE_UNIPOLAR_CHOICES["colors"][:-1]
```
Comment (niklasmohrin) [inline evap/evaluation/templatetags/evaluation_filters.py:118]: Was this wrong before? Do we need a test for it?

File: evap/results/tools.py [hunk 3]
```diff
@@ -42,55 +45,76 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
...
+class _CountedRatingResult(RatingResult):
+    counts: tuple[int, ...]
+    zipped_choices: list[list[Real | StrOrPromise | str | Real]]
+
+    count_sum: int
+    average: float
+    approval_count: int
+    minus_balance_count: float
+
+    is_published = True
```
[the remark this answers is not in the document]
Comment (Kakadus) [reply evap/results/tools.py:110]: you're right. If the `is_published` property would have remained in `RatingResult` it would make a difference though

File: evap/results/templates/distribution_with_grade.html [hunk 4]
```diff
@@ -5,7 +5,7 @@
     <div class="distribution-bar {{ question_result.choices.css_class }}"
         {% if question_result.question.is_bipolar_likert_question %} style="left: {{ question_result.minus_balance_count|percentage_one_decimal:question_result.count_sum }}"{% endif %}
     >
-        {% with colors=question_result.choices|to_colors %}
+        {% with colors=question_result|to_colors %}
             {% for value, color in distribution|zip:colors %}
```
Comment (Kakadus) [inline evap/results/templates/distribution_with_grade.html:9]: to your other comment: before, distribution values (w/o NO_ANSWER) may have been zipped with the colors with NO_ANSWER. 
Now, I explicitly remove NO_ANSWER from the colors and zip them strictly. I don't think a test is necessary, because this is already covered and only does some narrowing on the zipable tuples

File: evap/evaluation/templatetags/evaluation_filters.py (hunk 2 above)
Comment (Kakadus) [reply evap/evaluation/templatetags/evaluation_filters.py:118]: see comment below in distribution_with_grade

File: evap/results/tools.py [hunk 5]
```diff
@@ -42,55 +45,75 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
...
+        if not is_published(self):
             return None
         return sum(self.counts)
 
     @property
-    def minus_balance_count(self) -> float:
+    def minus_balance_count(self) -> float | None:
         assert self.question.is_bipolar_likert_question
+        if not is_published(self):
+            return None
```
Comment (niklasmohrin) [inline evap/results/tools.py:79]: In the tests, this never happens. Should this be an assert instead? If not, why?

File: evap/results/tools.py [hunk 6]
```diff
@@ -42,55 +45,75 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
...
+    zipped_choices: list[list[Real | StrOrPromise | str | Real]]
+
+    count_sum: int
+    average: float
+    approval_count: int
+    minus_balance_count: float
+
+
+def is_published(result: RatingResult) -> TypeGuard[_CountedRatingResult]:
+    return result.counts is not None
```
Comment (niklasmohrin) [inline evap/results/tools.py:112]: I don't know, this all feels a bit brittle, we only have this situation because we are passing `None` for `answer_counters` if we don't want to publish them. We should probably refactor this (at some point) to have actual typing describe the state of the system and not work around it with `TypeGuard`. Why is this needed here again?

File: evap/results/tools.py [hunk 7]
```diff
@@ -42,55 +45,75 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
...
-        return self.is_published and any(count != 0 for count in self.counts)
+        return has_answers(self)
 
-    @property
-    def is_published(self) -> bool:
-        return self.counts is not None
+
+class _CountedRatingResult(RatingResult):
+    counts: tuple[int, int, int, int]
+    zipped_choices: list[list[Real | StrOrPromise | str | Real]]
```
Comment (niklasmohrin) [inline evap/results/tools.py:103]:
```suggestion
    zipped_choices: list[list[Real | StrOrPromise]]
```

File: evap/results/tools.py (hunk 5 above)
Comment (Kakadus) [reply evap/results/tools.py:79]: The check/assert is needed because `self.counts` and `self.count_sum` is accessed. I just followed the style of the other classes. We can use the assert here instead if it's better

File: evap/results/tools.py (hunk 6 above)
Comment (Kakadus) [reply evap/results/tools.py:112]: You are right, the typeguards are not the the actual target of this PR and are not required here

**Review by richardebeling — COMMENTED** [review]
Looking good so far, but will have to recheck changes to the Result classes once #2091 is merged

**Review by richardebeling — APPROVED** [review]

**Review by niklasmohrin — APPROVED** [review]
